net: improves addnode / m_added_nodes logic

sr-gi opened the pull request:
## Rationale

Currently, `addnode` has a couple of corner cases that allow it to either connect to the same peer more than once, hence wasting outbound connection slots, or add redundant information to `m_added_nodes`, hence making Bitcoin iterate through useless data on a regular basis.

### Connecting to the same node more than once

In general, connecting to the same node more than once is something we should try to prevent. Currently, this is possible via `addnode` in two different ways:

1. Calling `addnode` more than once in a short time period, using two equivalent but distinct addresses
2. Calling `addnode add` using an IP, and `addnode onetry` after with an address that resolved to the same IP

For the former, the issue boils down to `CConnman::ThreadOpenAddedConnections` calling `CConnman::GetAddedNodeInfo` once, and iterating over the result to open connections (`CConman::OpenNetworkConnection`) on the same loop for all addresses.`CConnman::ConnectNode` only checks a single address, at random, when resolving from a hostname, and uses it to check whether we are already connected to it.

An example to test this would be calling:

```
bitcoin-cli addnode "127.0.0.1:port" add
bitcoin-cli addnode "localhost:port" add
```

And check how it allows us to perform both connections some times, and some times it fails.

The latter boils down to the same issue, but takes advantage of `onetry` bypassing the `CConnman::ThreadOpenAddedConnections` logic and calling `CConnman::OpenNetworkConnection` straightaway. A way to test this would be:

```
bitcoin-cli addnode "127.0.0.1:port" add
bitcoin-cli addnode "localhost:port" onetry
```

### Adding the same peer with two different, yet equivalent, addresses

The current implementation of `addnode` is pretty naive when checking what data is added to `m_added_nodes`. Given the collection stores strings, the checks at `CConnman::AddNode()` basically check wether the exact provided string is already in the collection. If so, the data is rejected, otherwise, it is accepted. However, ips can be formatted in several ways that would bypass those checks. 

Two examples would be `127.0.0.1` being equal to `127.1` and `[::1]` being equal to `[0:0:0:0:0:0:0:1]`. Adding any pair of these will be allowed by the rpc command, and both will be reported as connected by `getaddednodeinfo`, given they map to the same `CService`. 

This is less severe than the previous issue, since even tough both nodes are reported as connected by `getaddednodeinfo`, there is only a single connection to them (as properly reported by `getpeerinfo`). However, this adds redundant data to `m_added_nodes`, which is undesirable.


### Parametrize `CConnman::GetAddedNodeInfo`
Finally, this PR also parametrizes `CConnman::GetAddedNodeInfo` so it returns either all added nodes info, or only info about the nodes we are **not** connected to. This method is used both for `rpc`, in `getaddednodeinfo`, in which we are reporting all data to the user, so the former applies, and to check what nodes we are not connected to, in `CConnman::ThreadOpenAddedConnections`, in which we are currently returning more data than needed and then actively filtering using `CService.fConnected()`

sr-gi committed:
net/rpc: Check all resolved addresses in ConnectNode rather than just one

The current `addnode` rpc command has some edge cases in where it is possible to
connect to the same node twice by combining ip and address requests. This can happen under two situations:

The two commands are run one right after each other, in which case they will be processed
under the same loop in `CConnman::ThreadOpenAddedConnections` without refreshing `vInfo`, so both
will go trough. An example of this would be:

```
bitcoin-cli addnode "localhost:port" add

```

A node is added by IP using `addnode "add"` while the other is added by name using
`addnode "onetry"` with an address that resolves to multiple IPs. In this case, we currently
only check one of the resolved IPs (picked at random), instead of all the resolved ones, meaning
this will only probabilistically fail/succeed. An example of this would be:

```
bitcoin-cli addnode "127.0.0.1:port" add
[...]
bitcoin-cli addnode "localhost:port" onetry
```

Both cases can be fixed by iterating over all resolved addresses in `CConnman::ConnectNode` instead
of picking one at random

sr-gi committed:
rpc: Prevents adding the same ip more than once when formatted differently

Currently it is possible to add the same node twice when formatting IPs in
different, yet equivalent, manner. This applies to both ipv4 and ipv6, e.g:

127.0.0.1 = 127.1 | [::1] = [0:0:0:0:0:0:0:1]

`addnode` will accept both and display both as connected (given they translate to
the same IP). This will not result in multiple connections to the same node, but
will report redundant info when querying `getaddednodeinfo` and populate `m_added_nodes`
with redundant data.

This can be avoided performing comparing the contents of `m_added_addr` and the address
to be added as `CServices` instead of as strings.

sr-gi committed:
net/rpc: Makes CConnman::GetAddedNodeInfo able to return only non-connected address on request

`CConnman::GetAddedNodeInfo` is used both to get a list of addresses to manually connect to
in `CConnman::ThreadOpenAddedConnections`, and to report about manually added connections in
`getaddednodeinfo`. In both cases, all addresses added to `m_added_nodes` are returned, however
the nodes we are already connected to are only relevant to the latter, in the former they are
actively discarded.

Parametrizes `CConnman::GetAddedNodeInfo` so we can ask for only addresses we are not connected to,
to avoid passing useless information around.

sr-gi commented:
I've purposely split this into three commits according to each of the fixes so they can be discussed, and considered independently. That way it'll be easier to remove any if the change does not reach enough consensus.

kevkevinpal commented:
would it make sense to add a functional test like the below under
```
ip_port = "localhost:[]".format(p2p_port='add')
self.nodes[0].addnode(node=ip_port, command='add')
```
in `test/functional/rpc_net.py` on line 212
because we shortly after assert that there is only 1 added node added

sr-gi commented:
> would it make sense to add a functional test like the below under
> 
> ```
> ip_port = "localhost:[]".format(p2p_port='add')
> self.nodes[0].addnode(node=ip_port, command='add')
> ```
> 
> in `test/functional/rpc_net.py` on line 212 because we shortly after assert that there is only 1 added node added

I don't think that'd work. You'll still be able to add two nodes that resolve to the same IP using `addnode` as long as they belong to different namespaces. Lookups are not performed. The difference though is that both connections won't be established. However, the connection is not being checked by the test. 

What could be added, however, is a check for `127.0.0.1` and `127.1`.

PS: This actually made me realize that the check for whether to add something to `m_added_nodes` was not good enough.

sr-gi commented:
Added a test to `test/functional/rpc_net.py` and fixed the check on `CConnman::AddNode`

kevkevinpal commented on test/functional/rpc_net.py:
```diff
@@ -57,14 +57,14 @@ def run_test(self):
+        # self.test_connection_count()
+        # self.test_getpeerinfo()
+        # self.test_getnettotals()
+        # self.test_getnetworkinfo()
         self.test_getaddednodeinfo()
-        self.test_service_flags()
-        self.test_getnodeaddresses()
-        self.test_addpeeraddress()
+        # self.test_service_flags()
+        # self.test_getnodeaddresses()
+        # self.test_addpeeraddress()
```
might want to uncomment these other tests

sr-gi replied:
Ups, my bad

vasild commented on src/net.cpp:
```diff
@@ -462,19 +462,22 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
     if (pszDest) {
         const std::vector<CService> resolved{Lookup(pszDest, default_port, fNameLookup && !HaveNameProxy(), 256)};
         if (!resolved.empty()) {
-            const CService& rnd{resolved[GetRand(resolved.size())]};
```
The old code would pick a random address from the list of resolved addresses. The new code would pick the last one. I think it is good to preserve the randomness. One way to achieve that would be to shuffle the result before iterating and picking the last one.

```cpp
if (!resolved.empty()) {
    Shuffle(resolved.begin(), resolved.end(), FastRandomContext{});
    for (const auto& r : resolved) {
```

or if the shuffle is perceived expensive for big lists, before the loop `auto to_pick = GetRand(resolved.size())` and inside the loop only assign `addrConnect` for the `to_pick`th element.

sr-gi replied:
I think `Shuffle` should do, given this picks on a list of resolved DNS lookups. Normally, that should be rather short, and if [this](https://tmblr.co/ZCK6wx1dJN3aZ) is still up to date, the max length should be 4095.

vasild commented on src/net.cpp:
```diff
@@ -2570,9 +2577,10 @@ std::vector<CAddress> CConnman::GetAddresses(CNode& requestor, size_t max_addres
 
 bool CConnman::AddNode(const std::string& strNode)
 {
+    CService resolved(LookupNumeric(strNode, Params().GetDefaultPort(strNode)));
```
In the commit message of 12e601c33e `rpc: Prevents adding the same ip more than once when formatted differently`:

> ... formatting IPs in aborts
    different,

vasild commented on src/net.cpp:
```diff
@@ -2570,9 +2577,10 @@ std::vector<CAddress> CConnman::GetAddresses(CNode& requestor, size_t max_addres
     LOCK(m_added_nodes_mutex);
     for (const std::string& it : m_added_nodes) {
-        if (strNode == it) return false;
+        if (strNode == it || (resolved != CService{} && resolved == LookupNumeric(it, Params().GetDefaultPort(it)))) return false;
```
nit: use `resolved.IsValid()` instead of `resolved != CService{}`.

vasild commented on src/net.cpp:
```diff
@@ -1919,7 +1922,7 @@ std::vector<CAddress> CConnman::GetCurrentBlockRelayOnlyConns() const
 }
 
-std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
+std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo(bool includeConnected) const
```
style (same in the declaration):
```suggestion
std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo(bool include_connected) const
```

vasild commented on src/net.cpp:
```diff
@@ -1979,21 +1988,19 @@ void CConnman::ThreadOpenAddedConnections()
     {
         CSemaphoreGrant grant(*semAddnode);
-        std::vector<AddedNodeInfo> vInfo = GetAddedNodeInfo();
+        std::vector<AddedNodeInfo> vInfo = GetAddedNodeInfo(false);
```
```suggestion
        std::vector<AddedNodeInfo> vInfo = GetAddedNodeInfo(/*include_connected=*/false);
```
and in other call sites too

vasild reviewed (commented):
Approach ACK a6fcf17f9dde56637bcfe824bb3def83e764b38c

sr-gi commented:
Addressed @vasild comments

vasild commented on src/net.cpp:
```diff
@@ -462,19 +462,22 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
+                LOCK(m_nodes_mutex);
+                CNode* pnode = FindNode(static_cast<CService>(addrConnect));
+                if (pnode) {
+                    LogPrintf("Failed to open new connection, already connected\n");
```
This is now:

```cpp
LogPrintf("Not opening a connection to %d, already connected to %s\n", pnode->addr.ToStringAddrPort(), addrConnect..ToStringAddrPort());
```

It will print two IP addresses, which are most likely equal. It would be more valuable to print the hostname: "... pool.bitcoinnodes.org, already connected to 1.2.3.4" which implies that 1.2.3.4 is the same as pool.bitcoinnodes.org.

```cpp
LogPrintf("Not opening a connection to %s, already connected to %s\n", pszDest, addrConnect..ToStringAddrPort());
```

sr-gi replied:
Addressed in [079c56c](https://github.com/bitcoin/bitcoin/pull/28155/commits/079c56cfe7ed150a158b619f04e6d47edd01ca15)

vasild reviewed (approved):
ACK d0e7e181840a6c76f2a3a5cb39ccc5653c0e62e0

Would be nice to print the hostname here: https://github.com/bitcoin/bitcoin/pull/28155#discussion_r1304482465

vasild reviewed (approved):
ACK fd8b1db9ca8fb5aa0678d36e79b76bd283cf8fed

vasild reviewed (approved):
ACK afa15ed2c741593f49d608d75565daa5e55c6353

mzumsande commented on src/net.cpp:
```diff
@@ -464,21 +464,25 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
+                LOCK(m_nodes_mutex);
+                CNode* pnode = FindNode(static_cast<CService>(addrConnect));
+                if (pnode) {
+                    LogPrintf("Not opening a connection to %d, already connected to %s\n", pszDest, addrConnect.ToStringAddrPort());
```
This should be `%s` not `%d` (not that it actually matters with `tinyformat`)

mzumsande commented on src/test/fuzz/connman.cpp:
```diff
@@ -121,7 +121,7 @@ FUZZ_TARGET(connman, .init = initialize_connman)
             });
     }
-    (void)connman.GetAddedNodeInfo();
+    (void)connman.GetAddedNodeInfo(/*include_connected=*/true);
```
could fuzz the added arg `(fuzzed_data_provider.ConsumeBool()`)

mzumsande reviewed (commented):
Code review ACK afa15ed2c741593f49d608d75565daa5e55c6353

sr-gi commented:
Addresses comments from @mzumsande

sr-gi commented:
I've been revisiting the patch introduced by this PR to follow up based on https://github.com/bitcoin/bitcoin/pull/28155#discussion_r1362677038, and I realized that the following piece of code (net.cpp L558-L569) may not be required anymore (and it should be removed here too if so):

```cpp
if (addrConnect.IsValid()) {
    [...]
} else if (pszDest && GetNameProxy(proxy)) {
    sock = CreateSock(proxy.proxy);
    if (!sock) {
        return nullptr;
    }
    std::string host;
    uint16_t port{default_port};
    SplitHostPort(std::string(pszDest), port, host);
    bool proxyConnectionFailed;
    connected = ConnectThroughProxy(proxy, host, port, *sock, nConnectTimeout,
                                    proxyConnectionFailed);
}
```

To my understanding, it cannot be the case that `!addrConnect.IsValid()` while `pszDest` is set, given we now check for its validity after re-assigning `addrConnect` in L474-L478. Therefore, any `addrConnect` is ensured to be valid provided `pszDest` is set.

Does this make sense @vasild @mzumsande?

mzumsande reviewed (commented):
re-ACK 3d1854cdf31a83f4daafd06a760e3355f3bcc547

> To my understanding, it cannot be the case that `!addrConnect.IsValid()` while `pszDest` is set, given we now check for its validity after re-assigning `addrConnect` in L474-L478. Therefore, any `addrConnect` is ensured to be valid provided `pszDest` is set.
> 
> Does this make sense @vasild @mzumsande?

I don't think that's correct. If we are connecting through a proxy, `Lookup()` could return no entries, `addrConnect` would be default-initialized (and therefore invalid) and we might need this code.

sr-gi commented:
> ```c++
>     bool proxyConnectionFailed;
> ```

You're right!

sr-gi commented:
> re-ACK [3d1854c](https://github.com/bitcoin/bitcoin/commit/3d1854cdf31a83f4daafd06a760e3355f3bcc547)
> 
> > To my understanding, it cannot be the case that `!addrConnect.IsValid()` while `pszDest` is set, given we now check for its validity after re-assigning `addrConnect` in L474-L478. Therefore, any `addrConnect` is ensured to be valid provided `pszDest` is set.
> > Does this make sense @vasild @mzumsande?
> 
> I don't think that's correct. If we are connecting through a proxy, `Lookup()` could return no entries, `addrConnect` would be default-initialized (and therefore invalid) and we might need this code.

You're right!

jonatack reviewed (commented):
utACK 3d1854cdf31a83f4daafd06a760e3355f3bcc547 code review, still need to test these manually

Regarding `rpc: Prevents adding the same ip more than once when formatted differently`, I noticed this issue with I2P and CJDNS peers addnoded both with and without a port will cause multiple entries to be returned by RPC `getaddednodesinfo`. It looks like this commit may fix that and I'll test it manually. If yes, that might be a good test case to add. 

Commit `net/rpc: prevents connecting to an address we are already connected to` might be more accurately named `Check all resolved addresses in ConnectNode rather than one`, as there remain a number of open issues to solve related to peer connection detection, i.e. #28248.

jonatack committed:
Create net_peer_connection unit tests

for initial partial unit test coverage of these CConnman class methods:

- AddNode()
- ConnectNode()
- GetAddedNodeInfo()
- AlreadyConnectedToAddress()
- ThreadOpenAddedConnections()

and of the GetAddedNodeInfo() call in RPC addnode.

jonatack reviewed (commented):
As part of writing unit test coverage for #28248, I've written unit test coverage for all three commits of this pull.  Feel free to pull in the last 2 commits in https://github.com/bitcoin/bitcoin/pull/28749.  Suggest running the following to test it:

```bash
./src/test/test_bitcoin -t net_peer_connection_tests -l message -- -printtoconsole=1
```

or

```
./src/test/test_bitcoin -t net_peer_connection_tests -l all -- -printtoconsole=1 
```

jonatack committed:
Allow unit tests to access additional CConnman members

that are otherwise private:
- CConnman::m_nodes
- CConnman::ConnectNodes()
- CConnman::AlreadyConnectedToAddress()

and update the #include headers per iwyu.

sr-gi commented:
Addresses review comments and add @jonatack's test commits

fanquake commented on src/test/net_peer_connection_tests.cpp:
```diff
@@ -0,0 +1,152 @@
+#include <util/chaintype.h>
+#include <version.h>
+
+#ifdef WIN32
```
This should be including `compat.h`, rather than introducing the same networking #ifdef here?

sr-gi replied:
Rebased

kashifs commented:
tACK [0420f9](https://github.com/bitcoin/bitcoin/commit/0420f99f429ce2382057e101859067f40de47be0)

mzumsande reviewed (commented):
Tested ACK 0420f99f429ce2382057e101859067f40de47be0

sr-gi commented:
> tACK [0420f9](https://github.com/bitcoin/bitcoin/commit/0420f99f429ce2382057e101859067f40de47be0)

Thanks for reviewing! Would you mind sharing what did you test and how?

@vasild would you mind re-acking this when you have some time?

kashifs commented:
> > tACK [0420f9](https://github.com/bitcoin/bitcoin/commit/0420f99f429ce2382057e101859067f40de47be0)
> 
> 
> 
> Thanks for reviewing! Would you mind sharing what did you test and how?
> 
> 
> 
> @vasild would you mind re-acking this when you have some time?

I pulled the branch, compiled from source on my Mac, and ran:

./src/test/test_bitcoin -t net_peer_connection_tests -l message -- -printtoconsole=1

and

`./src/test/test_bitcoin -t net_peer_connection_tests -l all -- -printtoconsole=1`

I also pored over every line of modified code in order to better understand the design.

Is there anything else that I should do that might be helpful?

sr-gi commented:
> > > tACK [0420f9](https://github.com/bitcoin/bitcoin/commit/0420f99f429ce2382057e101859067f40de47be0)
> > 
> > 
> > 
> > Thanks for reviewing! Would you mind sharing what did you test and how?
> > 
> > 
> > 
> > @vasild would you mind re-acking this when you have some time?
> 
> I pulled the branch, compiled from source on my Mac, and ran:
> 
> `./src/test/test_bitcoin -t net_peer_connection_tests -l message -- -printtoconsole=1`
> 
> and
> 
> `./src/test/test_bitcoin -t net_peer_connection_tests -l all -- -printtoconsole=1`
> 
> I also pored over every line of modified code in order to better understand the design.
> 
> Is there anything else that I should do that might be helpful?

Nope, that's good. Thanks again for reviewing and testing.

vasild commented on src/test/util/net.cpp:
```diff
@@ -98,6 +102,17 @@ bool ConnmanTestMsg::ReceiveMsgFrom(CNode& node, CSerializedNetMsg&& ser_msg) co
     return complete;
 }
 
+CNode* ConnmanTestMsg::ConnectNodePublic(PeerManager& peerman, const char* pszDest, ConnectionType conn_type)
+{
+    CNode* node = ConnectNode(CAddress{}, pszDest, /*fCountFailure=*/false, conn_type, /*use_v2transport=*/true);
+    if (!node) return nullptr;
+    node->SetCommonVersion(PROTOCOL_VERSION);
+    peerman.InitializeNode(*node, ServiceFlags(NODE_NETWORK | NODE_WITNESS));
+    node->fSuccessfullyConnected = true;
+    AddTestNode(*node);
+    return node;
+}
```
nit, naming: maybe a different name would have been better than `ConnectNodePublic()`, given that it contains extra logic on top of the private `ConnectNode()`.

vasild replied:
Dunno, maybe `ConnectNodeAndInitialize()`?

vasild reviewed (commented):
ACK 0420f99f429ce2382057e101859067f40de47be0
